### DISCUSSION TITLE
Mongo DB Local Docker Replica Set - No Servers Available

### DISCUSSION THREAD

**Discussion by sloan-dog** [prisma/prisma#18215] [Q&A] [answered]
### Bug description

I am running mongo 6 in a replica set of 1 instance locally in docker.

I am able to connect to the replica set via mongosh, however, prisma yields this error, throw the client it generates and via the cli
```
Raw query failed. Code: `unknown`. Message: `Server selection timeout: No available servers. Topology: { Type: ReplicaSetNoPrimary, Servers: [ { Address: localhost:27017, Type: Unknown, Error: Connection refused (os error 61) }, ] }`
```

However, I can see via docker logs that prisma is connecting.

```
{"t":{"$date":"2023-03-05T04:39:25.298+00:00"},"s":"I",  "c":"NETWORK",  "id":22943,   "ctx":"listener","msg":"Connection accepted","attr":{"remote":"172.30.0.1:55334","uuid":"f9c33a70-6f26-4ce3-8c02-47539a4a46d4","connectionId":23,"connectionCount":1}}
{"t":{"$date":"2023-03-05T04:39:25.299+00:00"},"s":"I",  "c":"NETWORK",  "id":51800,   "ctx":"conn23","msg":"client metadata","attr":{"remote":"172.30.0.1:55334","client":"conn23","doc":{"driver":{"name":"mongo-rust-driver|Prisma","version":"2.3.0"},"os":{"type":"macos","architecture":"x86_64","version":"12.4.0"},"platform":"rustc 1.67.1 (d5a82bbd2 2023-02-07) with tokio"}}}
{"t":{"$date":"2023-03-05T04:39:55.518+00:00"},"s":"I",  "c":"NETWORK",  "id":22944,   "ctx":"conn23","msg":"Connection ended","attr":{"remote":"172.30.0.1:55334","uuid":"f9c33a70-6f26-4ce3-8c02-47539a4a46d4","connectionId":23,"connectionCount":0}}
```

I have verified that the mongo replica set is initialized and that the single node is in fact the primary. I don't have a clue as to why prisma can't negotiate connecting to primary. 

### Prisma information

```prisma
// This is your Prisma schema file,
// learn more about it in the docs: https://pris.ly/d/prisma-schema

generator client {
  provider = "prisma-client-js"
}

datasource db {
  provider = "mongodb"
  url      = env("DB_URL")
}
```

### Environment & setup

- OS: Mac OS Monterry
- Database: MongoDB 6
- Node.js version: 16.18.1

### Prisma Version

```
prisma                  : 4.11.0
@prisma/client          : 4.11.0
Current platform        : darwin
Query Engine (Node-API) : libquery-engine 8fde8fef4033376662cad983758335009d522acb (at node_modules/@prisma/engines/libquery_engine-darwin.dylib.node)
Migration Engine        : migration-engine-cli 8fde8fef4033376662cad983758335009d522acb (at node_modules/@prisma/engines/migration-engine-darwin)
Format Wasm             : @prisma/prisma-fmt-wasm 4.11.0-57.8fde8fef4033376662cad983758335009d522acb
Default Engines Hash    : 8fde8fef4033376662cad983758335009d522acb
Studio                  : 0.483.0
```

**Comment by sloan-dog** [comment, marked as answer]
**Solution**: Adding `directConnection=true` quey param to connection string allowed prisma's client to succesfully connect to the database. Without this, it failed to resolve. 

I also changed my init script to use a default by specifiying no conifg `rs.initiate()`. 

I hope this helps someone not spend the 6 hours of debugging this took before I started marrying random strings I saw in logs from mongosh to the db url...

**Steps to launch mongo in docker**

```
docker pull mongo:6

docker network create my-mongo-cluster

docker run \
-d \
-p 127.0.0.1:8232:27017 \
-e MONGO_INITDB_ROOT_USERNAME=$DB_USER \
-e MONGO_INITDB_ROOT_PASSWORD=$DB_PASSWORD \
--name mongo1 \
-v $(pwd)/key.txt:/keyfile \
--network my-mongo-cluster \
mongo:6 \
mongod --replSet myReplicaSet --bind_ip localhost,mongo1 --keyFile "/keyfile"
```

(key.txt is just some file containing a string)

then initialize with 
```
docker exec -it mongo1 mongosh --username $DB_USER --password $DB_PASSWORD --authenticationDatabase "admin" --eval "rs.initiate()"
```
**Connection in prisma**
Then if the prisma db url lives in an env var -> `mongodb://$DB_USER:$DB_PASSWORD@$DB_HOST:$DB_PORT/$DB_NAME?directConnection=true&serverSelectionTimeoutMS=2000&authSource=admin"`, if you have a hardcoded url (😨 ) then interpolate those variables ahead of time

Be sure to regenerate your client, if not connecting via CLI or prisma studio

**More information**
For more information about connecting to mongo replica sets -> https://www.mongodb.com/docs/mongodb-shell/connect/
(I found this after solving it, but helped me get slightly more clarity on what the problem was) 
TLDR -> mongosh appends the directionConnection string automatically if connecting to a specific replica set node. Prisma and/or its rust mongo client, do not.

**Reply by LeungKaMing** [reply]
It helps me a lot after I spent over a week to search issues or discussions, you save me!!! Hope someone who met this annoying and frustrating case that can see this answer.

**Comment by toritsejuFO** [comment]
You are a life saver. Thank you so much. Such a little details, hard to find.
